File: src/wiredtiger/table.cc
#include <wiredtiger.h>
#include "helpers.h"
#include "../helpers.h"
#include "db.h"
#include "table.h"
#include "session.h"
#include "multi_cursor.h"

#include <vector>

using namespace std;

namespace wiredtiger {
  WiredTigerTable::WiredTigerTable(WiredTigerDB *_db, char* _tableName, char* _config):
    tableConfig(NULL),
    tableName(NULL),
    db(_db)
  {
    if (_tableName != NULL) {
      tableName = (char*)calloc(sizeof(char), strlen(_tableName) + 1);
      strcpy(tableName, _tableName);
      specName = (char*)calloc(sizeof(char), strlen(_tableName) + 1 + 6); // table:
      sprintf(specName, "table:%s", tableName);
    }
    if (_config != NULL) {
      tableConfig = (char*)calloc(sizeof(char), strlen(_config) + 1);
      strcpy(tableConfig, _config);
    }
  }

  WiredTigerDB* WiredTigerTable::getDb() {
    return db;
  }

  char* WiredTigerTable::getTableName() {
    return tableName;
  }

  int WiredTigerTable::initTableFormats() {
    if (keyFormats.size() != 0) {
      return 0;
    }
    WT_CONFIG_PARSER* parser;
    RETURN_IF_ERROR(wiredtiger_config_parser_open(NULL, tableConfig, strlen(tableConfig), &parser));
    WT_CONFIG_ITEM item;
    RETURN_IF_ERROR(parser->get(parser, "key_format", &item));
    RETURN_IF_ERROR(parseFormat(item.str, item.len, &keyFormats));

    RETURN_IF_ERROR(parser->get(parser, "value_format", &item));
    RETURN_IF_ERROR(parseFormat(item.str, item.len, &valueFormats));
    return parser->close(parser);
  }



  int WiredTigerTable::initTable(WiredTigerSession* session) {
    if (isInitted) {
      return 0;
    }
    isInitted = true;
    initTableFormats();
    WiredTigerSession* actualSession = session;
    if (session == NULL) {
      WT_SESSION* wtSession;
      RETURN_IF_ERROR(db->conn->open_session(db->conn, NULL, NULL, &wtSession));
      actualSession = new WiredTigerSession(wtSession);
    }
    int result = actualSession->create(
      specName,
      tableConfig
    );
    if (session == NULL) {
      actualSession->close(NULL);
    }
    return result;
  }

  WiredTigerTable::~WiredTigerTable() {
    if (tableName != NULL) {
      free(tableName);
      free(specName);
    }
    if (tableConfig != NULL) {
      free(tableConfig);
    }
  }
  std::vector<Format>* WiredTigerTable::getKeyFormats() {
    initTableFormats();
    return &keyFormats;
  }
  std::vector<Format>* WiredTigerTable::getValueFormats() {
    initTableFormats();
    return &valueFormats;
  }

  int WiredTigerTable::createIndex(WiredTigerSession* session, char* indexName, char* config) {
    this->initTable(session);
    char* uri = (char*)calloc(sizeof(char), strlen(tableName) + strlen(indexName) + 8);
    sprintf(uri, "index:%s:%s", tableName, indexName);
    RETURN_AND_FREE_IF_ERROR(session->create(uri, config), uri);
    free(uri);
    return 0;
  }

  int WiredTigerTable::deleteMany(
    WiredTigerSession* session,
    std::vector<QueryCondition>* conditions,
    int* deletedCount
  ) {
    this->initTable(session);
    this->ensureSpecName();
    Cursor* deleteCursor;
    session->openCursor(specName, &deleteCursor);
    MultiCursor findCursor(session, tableName, conditions);
    int error;
    std::vector<unique_ptr<std::vector<QueryValue>>> keys;
    std::vector<QueryValue>* keyArray = NULL;
    while ((error = findCursor.next(&keyArray)) == 0) {
      keys.push_back(unique_ptr<std::vector<QueryValue>>(keyArray));
      deleteCursor->setKey(keyArray);
      int error = deleteCursor->remove();
      if (error == 0) {
        *deletedCount+=1;
      }
      if (error != 0 && error != WT_NOTFOUND) {
        delete deleteCursor;
        return error;
      }
    }
    delete deleteCursor;
    return 0;
  }

  int WiredTigerTable::insertMany(WiredTigerSession* session, std::vector<std::unique_ptr<EntryOfVectors>> *documents) {
    this->initTable(session);
    this->ensureSpecName();
    Cursor* cursor;
    RETURN_IF_ERROR(session->openCursor(specName, (char*)"overwrite=false", &cursor));
    // WT_CURSOR* cursor = wtCursor;
    for(size_t i = 0; i < documents->size(); i++) {
      EntryOfVectors& document = *documents->at(i);
      cursor->setKey(document.keyArray);
      cursor->setValue(document.valueArray);
      RETURN_AND_DELETE_IF_ERROR(cursor->insert(), cursor);
    }
    delete cursor;
    return 0;
  }

  int WiredTigerTable::updateMany(
    WiredTigerSession* session,
    std::vector<QueryCondition>* conditions,
    std::vector<QueryValue>* newValues,
    int* updatedCount
  ) {
    this->initTable(session);
    this->ensureSpecName();
    Cursor* updateCursor;
    RETURN_IF_ERROR(session->openCursor(specName, (char*)"overwrite=false", &updateCursor));
    MultiCursor findCursor(session, tableName, conditions);
    if (findCursor.error != 0) {
      return findCursor.error;
    }
    int error;
    // we can't delete keys until we're done, because we use them to dedupe overlapping index bounds
    std::vector<unique_ptr<std::vector<QueryValue>>> keys;
    std::vector<QueryValue>* keyArray = NULL;
    std::vector<QueryValue>* valueArray = NULL;
    int i = 0;
    while ((error = findCursor.next(&keyArray, &valueArray)) == 0) {
      keys.push_back(unique_ptr<std::vector<QueryValue>>(keyArray));
      updateCursor->setKey(keyArray);
      for (size_t i = 0; i < valueArray->size(); i++) {
        if ((*newValues)[i].dataType != FIELD_PADDING) {
          (*valueArray)[i].value.valuePtr = (*newValues)[i].value.valuePtr;
          (*valueArray)[i].value.valueUint = (*newValues)[i].value.valueUint;
          (*valueArray)[i].size = (*newValues)[i].size;
        }
      }
      updateCursor->setValue(valueArray);
      int error = updateCursor->update();
      if (error == 0) {
        *updatedCount+=1;
      }
      delete valueArray;
      if (error != 0 && error != WT_NOTFOUND) {
        delete updateCursor;
        return error;
      }
      RETURN_AND_DELETE_IF_ERROR(updateCursor->reset(), updateCursor);
    }
    delete updateCursor;
    return 0;
  }
}

File: src/glue/table.rs


use std::collections::HashMap;
use std::ffi::CString;


use super::cursor_trait::InternalCursorTrait;
use super::find_cursor::InternalFindCursor;
use super::multi_cursor::MultiCursor;
use super::query_value::{Format, InternalDocument, InternalIndexSpec, QueryValue};
use super::session::InternalSession;
use super::utils::extract_formats_from_config;
use super::error::*;

pub struct InternalTable {
  table_name: String,
  config: String,
  spec_name: String,
  is_initted: bool,

  key_formats: Vec<Format>,
  value_formats: Vec<Format>,
  pub index_formats: HashMap<String, Vec<Format>>
}

impl InternalTable {
  pub fn new(table_name: String, config: String) -> Self {
    let spec_name = format!("table:{}", table_name);
    return InternalTable {
      table_name,
      config,
      spec_name,
      is_initted: false,
      key_formats: Vec::new(),
      value_formats: Vec::new(),
      index_formats: HashMap::new()
    };
  }

  pub fn init_table_formats(&mut self) -> Result<(), GlueError> {
    if self.key_formats.len() != 0 {
      return Ok(());
    }
    return extract_formats_from_config(self.config.to_string(), &mut self.key_formats, &mut self.value_formats);
  }

  fn init_table(&mut self, session: &InternalSession) -> Result<(), GlueError> {
    if self.is_initted {
      return Ok(());
    }

    self.init_table_formats()?;
    session.create(self.spec_name.to_string(), Some(self.config.to_string()))?;

    Ok(())
  }

  pub fn get_key_formats(&self) -> Result<&Vec<Format>, GlueError> {
    Ok(&self.key_formats)
  }

  pub fn get_index_formats(&self) -> Result<&HashMap<String, Vec<Format>>, GlueError> {
    Ok(&self.index_formats)
  }

  // fuck this method - it's needed because of the "can't borrow immutable and mutable"
  pub fn get_index_and_key_formats(&mut self) -> Result<(&Vec<Format>, &HashMap<String, Vec<Format>>), GlueError> {
    self.init_table_formats()?;
    Ok((&self.key_formats, self.get_index_formats()?))
  }

  pub fn get_value_formats(&self) -> Result<&Vec<Format>, GlueError> {
    Ok(&self.value_formats)
  }

  pub fn insert_many(&mut self, session: &InternalSession, documents: &mut Vec<InternalDocument>) -> Result<i32, GlueError> {
    self.init_table(session)?;
    let len = documents.len();
    let mut cursor = session.open_cursor(self.spec_name.to_string(), None)?;
    for document in documents {
      cursor.set_key(&mut document.key)?;
      cursor.set_value(&mut document.value)?;
      cursor.insert()?;
    }
    (&mut cursor).close()?;
    Ok(len as i32)
  }

  // this function ensures we've loaded all the indexes created outside of create_index (just in case)
  pub fn ensure_index_formats(&mut self, session: &InternalSession, index_uris: &Vec<String>) -> Result<(), GlueError> {
    let metadata_cursor = session.open_cursor("metadata:create".to_string(), None)?;
    let mut qvs:Vec<QueryValue> = Vec::with_capacity(1);
    qvs.push(QueryValue::empty());
    for index_uri in index_uris {
      if self.index_formats.contains_key(index_uri) {
        continue;
      }
      let qv = qvs.get_mut(0).unwrap();
      qv.data_type = 'S';
      qv.set_external_str_box(CString::new(index_uri.to_string()).unwrap())?;
      metadata_cursor.set_key(&mut qvs)?;
      metadata_cursor.search()?;
      let index_config = metadata_cursor.get_value()?.get(0).unwrap().get_str()?.to_str().unwrap().to_string();
      let mut key_formats:Vec<Format> = Vec::new();
      let mut value_formats:Vec<Format> = Vec::new();
      extract_formats_from_config(index_config, &mut key_formats, &mut value_formats)?;
      self.index_formats.insert(index_uri.to_string(), key_formats);
    }
    Ok(())
  }

  pub fn create_index(&mut self, session: &InternalSession, index_name: String, config: String) -> Result<(), GlueError> {
    self.init_table(&session)?;
    session.create(format!("index:{}:{}", self.table_name, index_name), Some(config.clone()))?;

    let mut key_formats:Vec<Format> = Vec::new();
    let mut value_formats:Vec<Format> = Vec::new();
    // TODO: This only works if you specify the key_format - otherwise we need a way to extract format from column names
    extract_formats_from_config(config, &mut key_formats, &mut value_formats)?;

    self.index_formats.insert(format!("index:{}:{}", self.table_name, index_name), key_formats);
    Ok(())
  }

  pub fn find(
    &self,
    session: &InternalSession,
    index_specs: Vec<InternalIndexSpec>,
    columns_joined: Option<String>
  ) -> Result<InternalFindCursor, GlueError> {
    let mut cursor = InternalFindCursor::new(self.table_name.clone(), index_specs, columns_joined)?;
    cursor.init(session)?;
    return Ok(cursor);
  }

  pub fn delete_many(&self, session: InternalSession, index_specs: Vec<InternalIndexSpec>) -> Result<i32, GlueError> {
    let mut cursor = MultiCursor::new(self.table_name.clone(), index_specs, Some("()".to_string()))?;
    cursor.init(&session)?;
    let delete_cursor = session.open_cursor(format!("table:{}", self.table_name), None)?;
    let mut count = 0;
    while cursor.next()? {
      let mut key = cursor.get_key()?;
      for k in key.iter_mut() {
        k.set_external_reference()?;
      }
      delete_cursor.set_key(&mut key)?;
      delete_cursor.remove()?;
      count += 1;
    }

    return Ok(count);
  }

  pub fn update_many(&mut self, session: InternalSession, index_specs: Vec<InternalIndexSpec>, value: &mut Vec<QueryValue>) -> Result<i32, GlueError> {
    self.init_table(&session)?;
    let mut cursor = MultiCursor::new(self.table_name.clone(), index_specs, None)?;
    cursor.init(&session)?;
    let mut update_cursor = session.open_cursor(format!("table:{}", self.table_name), None)?;
    let mut count = 0;
    while cursor.next()? {
      let mut key = cursor.get_key()?;
      let existing = cursor.get_value()?;
      for k in key.iter_mut() {
        k.set_external_reference()?;
      }
      update_cursor.set_key(&mut key)?;

      for i in  0..value.len() {
        let value_in = value.get_mut(i).unwrap();
        let value_existing = existing.get(i).unwrap();
        if value_in.data_type == 'x' {
          value_in.internal_value = value_existing.internal_value.clone();
          value_in.set_external_reference()?;
        }
      }
      update_cursor.set_value(value)?;
      update_cursor.update()?;
      count += 1;
    }
    update_cursor.close()?;
    cursor.close()?;
    Ok(count)
  }
}
